dags/springer/parser.py

```diff
@@ -0,0 +1,259 @@
+
+        result = [
+            node.text
+            for node in [
+                org_div_node,
+                org_name_node,
+                street_node,
+                city_node,
+                state_node,
+                postcode_node,
+                country_node,
+            ]
+            if node is not None
+        ]
+
+        return ", ".join(result), org_name_node.text, country_node.text
+
+    def _get_published_date(self, article: ET.Element):
+        year = article.find(
+            "./Journal/Volume/Issue/Article/ArticleInfo/*/OnlineDate/Year"
+        ).text
+        month = article.find(
+            "./Journal/Volume/Issue/Article/ArticleInfo/*/OnlineDate/Month"
+        ).text
+        day = article.find(
+            "./Journal/Volume/Issue/Article/ArticleInfo/*/OnlineDate/Day"
+        ).text
+        return datetime.date(day=int(day), month=int(month), year=int(year)).isoformat()
+
+    def _get_affiliations(self, author_group: ET.Element, contrib: ET.Element):
+        affiliations = []
+        referred_id = contrib.get("AffiliationIDS")
+
+        if not referred_id:
+            return affiliations
+
+        for ref in referred_id.split():
+            cleaned_aff = self._clean_aff(
+                author_group.find(f"./Affiliation[@ID='{ref}']")
+            )
+            if cleaned_aff not in affiliations:
+                affiliations.append(cleaned_aff)
+
+        mapped_affiliations = [
+            {"value": aff, "organization": org, "country": country}
+            for aff, org, country, in affiliations
+        ]
+
+        return mapped_affiliations
+
+    def _get_authors(self, article: ET.Element):
+        authors = []
+        for contrib in article.findall(
+            "./Journal/Volume/Issue/Article/ArticleHeader/AuthorGroup/Author"
+        ):
+            author = ObjectExtractor(
+                None,
+                [
+                    TextExtractor("surname", "./AuthorName/FamilyName"),
+                    TextExtractor("given_names", "./AuthorName/GivenName"),
+                    TextExtractor("email", "./Contact/Email", False),
+                ],
+            ).extract(contrib)
+            author["affiliations"] = self._get_affiliations(
+                article.find(
+                    "./Journal/Volume/Issue/Article/ArticleHeader/AuthorGroup"
+                ),
+                contrib,
+            )
+
+            authors.append(author)
+
+        # if not authors:
+        #     logger.error('No authors found for article %s.' % dois)
+
+        return authors
+
+    def _get_page_nrs(self, article: ET.Element):
+        first_page = article.find(
+            "./Journal/Volume/Issue/Article/ArticleInfo/ArticleFirstPage"
+        ).text
+        last_page = article.find(
+            "./Journal/Volume/Issue/Article/ArticleInfo/ArticleLastPage"
+        ).text
+        return [int(last_page) - int(first_page) + 1]
```

Comment:
what if last_page or first_page is empty string?

Reply:
Can this happen ? If so, how should we process them ? Currently it will just throw an exception, as it should Imo

Reply:
It's not a required field that's why I would just log and not fail the whole harvest.

Reply:
About that, I currently don't have proper logging, should I add it now ? On your leftover comments I removed the comments, but maybe I should just add the logging
$

Reply:
Yes we should logging, could you make a separate PR? We can use https://github.com/hynek/structlog to give structure in the logs having information like `doi`, `file` to easily filter logs